src/OpenTelemetry.Api/Context/Propagation/AWSXRayPropagator.cs

```diff
@@ -0,0 +1,302 @@
+
+        /// <inheritdoc/>
+        public PropagationContext Extract<T>(PropagationContext context, T carrier, Func<T, string, IEnumerable<string>> getter)
+        {
+            if (context.ActivityContext.IsValid())
+            {
+                return context;
+            }
+
+            if (carrier == null)
+            {
+                OpenTelemetryApiEventSource.Log.FailedToExtractActivityContext(nameof(AWSXRayPropagator), "null carrier");
+                return context;
+            }
+
+            if (getter == null)
+            {
+                OpenTelemetryApiEventSource.Log.FailedToExtractActivityContext(nameof(AWSXRayPropagator), "null getter");
+                return context;
+            }
+
+            try
+            {
+                var parentTraceHeader = getter(carrier, AWSXRayTraceHeaderKey);
+
+                if (parentTraceHeader == null || parentTraceHeader.Count() != 1)
+                {
+                    return context;
+                }
+
+                var parentHeader = parentTraceHeader.First();
+
+                if (!TryParseXRayTraceHeader(parentHeader, out var newActivityContext))
+                {
+                    return context;
+                }
+
+                return new PropagationContext(newActivityContext, context.Baggage);
+            }
+            catch (Exception ex)
+            {
+                OpenTelemetryApiEventSource.Log.ActivityContextExtractException(nameof(AWSXRayPropagator), ex);
+            }
+
+            return context;
+        }
+
+        /// <inheritdoc/>
+        public void Inject<T>(PropagationContext context, T carrier, Action<T, string, string> setter)
+        {
+            if (context.ActivityContext.TraceId == default || context.ActivityContext.SpanId == default)
+            {
+                OpenTelemetryApiEventSource.Log.FailedToInjectActivityContext(nameof(AWSXRayPropagator), "Invalid context");
+                return;
+            }
+
+            if (carrier == null)
+            {
+                OpenTelemetryApiEventSource.Log.FailedToInjectActivityContext(nameof(AWSXRayPropagator), "null carrier");
+                return;
+            }
+
+            if (setter == null)
+            {
+                OpenTelemetryApiEventSource.Log.FailedToInjectActivityContext(nameof(AWSXRayPropagator), "null setter");
+                return;
+            }
+
+            var traceId = ToXRayTraceIdFormat(context.ActivityContext.TraceId.ToHexString());
+            var spanId = context.ActivityContext.SpanId.ToHexString();
+            var sampleDecision = (context.ActivityContext.TraceFlags & ActivityTraceFlags.Recorded) != 0 ? SampledValue : NotSampledValue;
+
+            var newTraceHeader = string.Concat(RootKey, KeyValueDelimiter, traceId, TraceHeaderDelimiter, ParentKey, KeyValueDelimiter, spanId, TraceHeaderDelimiter, SampledKey, KeyValueDelimiter, sampleDecision);
```

Comment:
This string.Concat could be brutal on perf. An option to consider: https://docs.microsoft.com/en-us/dotnet/api/system.string.create?view=netcore-3.1

Let's you fill in the string as it is constructed to reduce copying, more or less.

Only available to netstandard2.1/netcoreapp3.0+ though so we would have to add the netstandard2.1 target and a preprocessor region. Worth it, IMO.

Reply:
Thanks, will take a look and update